hivemind/p2p/servicer.py

```diff
@@ -98,25 +109,41 @@ async def caller(
                 self: StubBase, input: input_type, timeout: Optional[float] = None
             ) -> handler.response_type:
                 return await asyncio.wait_for(
-                    self._p2p.call_protobuf_handler(self._peer, handler.handle_name, input, handler.response_type),
+                    self._p2p.call_protobuf_handler(
+                        self._peer,
+                        cls._get_handle_name(self._namespace, handler.method_name),
+                        input,
+                        handler.response_type,
+                    ),
                     timeout=timeout,
                 )
 
         caller.__name__ = handler.method_name
         return caller
 
-    async def add_p2p_handlers(self, p2p: P2P, wrapper: Any = None) -> None:
+    async def add_p2p_handlers(self, p2p: P2P, wrapper: Any = None, *, namespace: Optional[str] = None) -> None:
+        self._collect_rpc_handlers()
+
         servicer = self if wrapper is None else wrapper
         for handler in self._rpc_handlers:
             await p2p.add_protobuf_handler(
-                handler.handle_name,
+                self._get_handle_name(namespace, handler.method_name),
                 getattr(servicer, handler.method_name),
                 handler.request_type,
                 stream_input=handler.stream_input,
             )
 
-    def get_stub(self, p2p: P2P, peer: PeerID) -> StubBase:
-        return self._stub_type(p2p, peer)
+    @classmethod
+    def get_stub(cls, p2p: P2P, peer: PeerID, *, namespace: Optional[str] = None) -> StubBase:
```

Comment:
Can you please explain the purpose of namespaces for RPCs? Are they used to avoid possible collisions for remote calls?

Reply:
By default, RPC handlers are named as `ClassName.method_name`. This does not allow to use two averagers of the same class that work with different tensors and parameters, however @justheuristic requested this functionality for experiments.

Defining `namespace` changes the handler name to `namespace::ClassName.method_name`, therefore one can serve two sets of averager RPCs (for two averagers) at one endpoint.